[SPARK-19634][ML] Multivariate summarizer - dataframes API

thunterdb opened the pull request:
## What changes were proposed in this pull request?

This patch adds the DataFrames API to the multivariate summarizer (mean, variance, etc.). In addition to all the features of `MultivariateOnlineSummarizer`, it also allows the user to select a subset of the metrics. This should resolve some performance issues related to computing unrequested metrics.

Furthermore, it uses the BLAS API to the extent possible, so that the given code should be efficient for the dense case.

## How was this patch tested?

This patch includes most of the tests of the RDD-based. It compares results against the existing `MultivariateOnlineSummarizer` as well as adding more tests.

This patch also includes some documentation for some low-level constructs such as `TypedImperativeAggregate`.

## Performance

I have not run tests against the existing implementation. However, this patch uses the recommended low-level SQL APIs, so it should be interesting to compare both implementation in that respect.

Thanks to @hvanhovell and Cheng Liang for suggestions on SparkSQL.

thunterdb committed:
work

thunterdb committed:
work on the test suite

thunterdb committed:
last work

thunterdb committed:
work on using imperative aggregators

thunterdb committed:
Merge remote-tracking branch 'upstream/master' into 19634

thunterdb committed:
more work on summarizer

thunterdb committed:
work

thunterdb committed:
changes

thunterdb committed:
Merge remote-tracking branch 'upstream/master' into 19634

thunterdb committed:
cleanup

thunterdb committed:
debugging

thunterdb committed:
work

thunterdb committed:
Merge remote-tracking branch 'upstream/master' into 19634

thunterdb committed:
debug

thunterdb committed:
trying to debug serialization issue

thunterdb committed:
better tests

thunterdb committed:
changes

thunterdb committed:
debugging

thunterdb committed:
more tests and debugging

thunterdb committed:
fixed tests

thunterdb committed:
doc

SparkQA commented:
**[Test build #75187 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75187/testReport)** for PR 17419 at commit [`35eaeb0`](https://github.com/apache/spark/commit/35eaeb0d02ae9cc29ae559231fe4858935315477).

thunterdb committed:
cleanups

thunterdb committed:
cleanups

SparkQA commented:
**[Test build #75188 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75188/testReport)** for PR 17419 at commit [`58b17dc`](https://github.com/apache/spark/commit/58b17dc3af5ebc5eaeff82559bc3be3a1f64ef38).

thunterdb committed:
Cleanups

SparkQA commented:
**[Test build #75189 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75189/testReport)** for PR 17419 at commit [`ffe5cfe`](https://github.com/apache/spark/commit/ffe5cfe2e5a60f3f6056d8f89f7c50bbedca3fbf).

thunterdb committed:
Cleanups

thunterdb committed:
Cleanups

SparkQA commented:
**[Test build #75190 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75190/testReport)** for PR 17419 at commit [`ba200bb`](https://github.com/apache/spark/commit/ba200bb1d52d468acebc9fe44f6727c570a35335).

SparkQA commented:
**[Test build #75188 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75188/testReport)** for PR 17419 at commit [`58b17dc`](https://github.com/apache/spark/commit/58b17dc3af5ebc5eaeff82559bc3be3a1f64ef38).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75188/
Test FAILed.

SparkQA commented:
**[Test build #75187 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75187/testReport)** for PR 17419 at commit [`35eaeb0`](https://github.com/apache/spark/commit/35eaeb0d02ae9cc29ae559231fe4858935315477).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75187/
Test FAILed.

SparkQA commented:
**[Test build #75189 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75189/testReport)** for PR 17419 at commit [`ffe5cfe`](https://github.com/apache/spark/commit/ffe5cfe2e5a60f3f6056d8f89f7c50bbedca3fbf).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75189/
Test FAILed.

SparkQA commented:
**[Test build #75190 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75190/testReport)** for PR 17419 at commit [`ba200bb`](https://github.com/apache/spark/commit/ba200bb1d52d468acebc9fe44f6727c570a35335).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75190/
Test FAILed.

SparkQA commented:
**[Test build #3614 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3614/testReport)** for PR 17419 at commit [`ba200bb`](https://github.com/apache/spark/commit/ba200bb1d52d468acebc9fe44f6727c570a35335).

SparkQA commented:
**[Test build #3614 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3614/testReport)** for PR 17419 at commit [`ba200bb`](https://github.com/apache/spark/commit/ba200bb1d52d468acebc9fe44f6727c570a35335).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

jkbradley commented:
I'll take a look later today.  Also CCing @dbtsai I know you worked on the old summarizer, so I thought I'd ping you here.  + @yanboliang and @holdenk since you commented on the design doc.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala:
```diff
@@ -0,0 +1,746 @@
+ * [[Summarizer]].
+ */
+@Since("2.2.0")
+abstract class SummaryBuilder {
```
Make this sealed
Also mark Experimental

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala:
```diff
@@ -0,0 +1,746 @@
+ * }}}
+ */
+@Since("2.2.0")
+object Summarizer extends Logging {
```
Mark Experimental

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala:
```diff
@@ -0,0 +1,746 @@
+    new SummaryBuilderImpl(typedMetrics, computeMetrics)
+  }
+
+  def mean(col: Column): Column = getSingleMetric(col, "mean")
```
Add Since annotations for all of these public methods

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala:
```diff
@@ -0,0 +1,338 @@
+  //
+  // The tests take an list of input vectors and a list of all the summary values that
+  // are expected for this input. They currently test against some fixed subset of the
+  // metrics, but should be made fuzzy in the future.
```
fuzzy?

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala:
```diff
@@ -0,0 +1,338 @@
+      case x => throw new Exception(x.toString)
+    }
+
+    val s = {
```
rename s -> summarizer

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala:
```diff
@@ -0,0 +1,338 @@
+    }
+
+    // Because the Spark context is reset between tests, we cannot hold a reference onto it.
+    def wrapped() = {
```
rename wrapped -> wrappedInit

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala:
```diff
@@ -0,0 +1,338 @@
+
+  private def compare(df: DataFrame, exp: Seq[Any]): Unit = {
+    val coll = df.collect().toSeq
+    val Seq(row) = coll
```
How about ```val res = df.first().toSeq```?

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala:
```diff
@@ -0,0 +1,338 @@
+  }
+
+  // Compares structured content.
+  private def compareStructures(x1: Any, x2: Any, name: String): Unit = (x1, x2) match {
```
Are there bugs in direct comparison of Rows?  I'm wondering if we can avoid this here.  Or, does Spark SQL already have this implemented?

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala:
```diff
@@ -0,0 +1,338 @@
+      assert(x, hint)
+    } catch {
+      case tfe: TestFailedException =>
+        throw new TestFailedException(Some(s"Failure with hint $hint"), Some(tfe), 1)
```
Why do you need to catch and re-throw it as a TestFailedException?

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala:
```diff
@@ -0,0 +1,338 @@
+    b
+  }
+
+  private def b(x: Array[Double]): Vector = Vectors.dense(x)
```
Rename these to something clearer

viirya replied:
+1

jkbradley reviewed (commented):
I just had time for some initial comments.  More later.

thunterdb committed:
Merge remote-tracking branch 'upstream/master' into 19634

SparkQA commented:
**[Test build #75268 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75268/testReport)** for PR 17419 at commit [`2f809ef`](https://github.com/apache/spark/commit/2f809efec0dfef8d94b747eec5f23bc61c9ff4cc).

dbtsai commented:
Will be really interested to see the performance benchmark durning the QA period so users can know when to use the dataframe apis or existing rdd apis. Thanks.

SparkQA commented:
**[Test build #75268 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75268/testReport)** for PR 17419 at commit [`2f809ef`](https://github.com/apache/spark/commit/2f809efec0dfef8d94b747eec5f23bc61c9ff4cc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75268/
Test PASSed.

sethah commented:
Is this being targeted for Spark 2.2?

thunterdb commented:
@sethah it would have been nice, but I do not think we should merge it this late into the release cycle.

thunterdb committed:
small test to find perf issues

SparkQA commented:
**[Test build #75285 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75285/testReport)** for PR 17419 at commit [`662f62c`](https://github.com/apache/spark/commit/662f62c5dc13b0b8b21f1c0fa374e4279f724a8e).

thunterdb commented:
I have added a small perf test to find the performance bottlenecks. Note that this test works on the worst case (vectors of size 1) from the perspective of overhead. Here are the numbers I currently get. I will profile the code to see if there are some obvious targets for optimization:

```
[min ~ median ~ max], higher is better:

RDD = [2482 ~ 46150 ~ 48354] records / milli
dataframe (variance only) = [4217 ~ 4557 ~ 4848] records / milli
dataframe = [2887 ~ 4420 ~ 4717] records / milli
```

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75285/
Test PASSed.

viirya commented:
> RDD = [2482 ~ 46150 ~ 48354] records / milli

the number is so varied?

kiszk commented on mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala:
```diff
@@ -335,4 +335,65 @@ class SummarizerSuite extends SparkFunSuite with MLlibTestSparkContext {
+    val rdd1 = sc.parallelize(1 to n).map { idx =>
+      OldVectors.dense(idx.toDouble)
+    }
+    val trieouts = 10
```
I think that 10 times without warmup is too small for performance measurement.
Can we use `Benchmark` class or add warmup run as `Benchmark` class does?

thunterdb replied:
I did not try about that class, thanks. It should stabilize the results.

viirya commented on mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala:
```diff
@@ -0,0 +1,746 @@
+private[ml]
+object SummaryBuilderImpl extends Logging {
+
+  def implementedMetrics: Seq[String] = allMetrics.map(_._1).sorted
```
`lazy val` should be enough.

viirya commented on mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala:
```diff
@@ -0,0 +1,399 @@
+
+  private def b(x: Array[Double]): Vector = Vectors.dense(x)
+
+  private def l(x: Array[Long]): Vector = b(x.map(_.toDouble))
```
And this. It is better to rename it.

viirya commented on mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala:
```diff
@@ -0,0 +1,746 @@
+      // All the fields that we compute on demand:
+      // TODO: the most common case is dense vectors. In that case we should
+      // directly use BLAS instructions instead of iterating through a scala iterator.
+      v.foreachActive { (index, value) =>
```
RDD's Summarizer doesn't have BLAS optimization actually. So this may not be the reason for the performance gap.

thunterdb replied:
Oh yes it does not. Note that the benchmark below is works with vectors of size 1, so as to analyze the overhead of dataframes vs RDDs. I will put a more realistic benchmark later.

viirya commented on mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala:
```diff
@@ -0,0 +1,746 @@
+    override def update(buff: Buffer, row: InternalRow): Buffer = {
+      // Unsafe rows do not play well with UDTs, it seems.
+      // Directly call the deserializer.
+      val v = udt.deserialize(row.getStruct(0, udt.sqlType.size))
```
This should hurt performance a lot.

When deserializing a vector, we copy (indices and) values array from the unsafe row.

WeichenXu123 replied:
I think
```
val v = udt.deserialize(row.getStruct(0, udt.sqlType.size))
```
has some problems.
We cannot directly use `getter` method on `row` parameter passed in, because the ordinal of the column we want to get depends on the underlying catalyst layer, instead, we should use:
```
val datum = child.eval(row) 
val featureVector  = udt.deserialize(datum)
```
to get column value.

WeichenXu123 replied:
If we want to use `weight` column when summarizing, I think we can define the UDAF as:
`summary(featureCol, weightCol)`
and in the constructor of `MetricsAggregate` pass the weight column in.
Example code as following:

```
case class MetricsAggregate(
      requested: Seq[Metrics],
      featureExpr: Expression, // feature column expr
      weightExpr: Expression, // weight column expr
      mutableAggBufferOffset: Int,
      inputAggBufferOffset: Int) extends TypedImperativeAggregate[Buffer] {

override def update(buff: Buffer, row: InternalRow): Buffer = {
      val featureVector = udt.deserialize(featureExpr.eval(row))
      val weight = weightExpr.eval(row)
      Buffer.updateInPlace(buff, featureVector, weight)
      buff
    }
}

def summary(featureCol: Column, weightCol: Column): Column = {
    val agg = MetricsAggregate(
      requestedMetrics,
      featureCol.expr,
      weightCol.expr,
      mutableAggBufferOffset = 0,
      inputAggBufferOffset = 0)
    new Column(AggregateExpression(agg, mode = Complete, isDistinct = false))
}

// handle the case user do not specify weight column
def summary(featureCol: Column): Column = {
    summary(featureCol, lit(1.0))
}

```
cc @cloud-fan @liancheng @yanboliang

yanboliang replied:
+1 @WeichenXu123.

thunterdb committed:
Current speed:

Dataframes = [2766.648008567718 ~ 5091.204527768661 ~ 5716.359795809639] records / milli

thunterdb committed:
BLAS calls for dense interface

SparkQA commented:
**[Test build #75406 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75406/testReport)** for PR 17419 at commit [`a569dac`](https://github.com/apache/spark/commit/a569dac8998d63adbc9adba1f2eb2f42967533e7).

thunterdb commented:
I looked a bit deeper into the performance aspect. Here are some quick insights:
 - there was an immediate bottleneck in `VectorUDT`, which boosts the performance already by 3x
 - it is not clear if switching to pure Breeze operations helps given the overhead for tiny vectors. I will need to do more analysis on larger vectors.
 - now, most of the time is roughly split between `ObjectAggregationIterator.processInputs` (40%), some codegen'ed expression (20%) and our own `MetricsAggregate.update` (35%)

That benchmark focuses on the overhead of catalyst. I will do another benchmark with dense vectors to see how it fares in practice with more real data.

SparkQA commented:
**[Test build #75406 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75406/testReport)** for PR 17419 at commit [`a569dac`](https://github.com/apache/spark/commit/a569dac8998d63adbc9adba1f2eb2f42967533e7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75406/
Test FAILed.

cloud-fan commented on mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala:
```diff
@@ -0,0 +1,799 @@
+
+  private case class MetricsAggregate(
+      requested: Seq[Metrics],
+      startBuffer: Buffer,
```
we should not pass around the `startBuffer`, but create an initial one in `createAggregationBuffer`

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala:
```diff
@@ -0,0 +1,799 @@
+    }
+
+    private def updateInPlaceDense(buffer: Buffer, v: DenseVector, w: Double): Unit = {
+      val epsi = Double.MinPositiveValue
```
Does the purpose of the code using breeze here to use BLAS to improve performance ?
BUT in breeze implementation ops between vectors do not use BLAS, instead in breeze it use `cForRange`.
cc @yanboliang

yanboliang replied:
I think the intention here is to make sequential operation conveniently and efficiently. AFAIK, ```cForRange``` is also very efficient. Thanks.

WeichenXu123 commented on mllib/src/test/scala/org/apache/spark/ml/stat/SummarizerSuite.scala:
```diff
@@ -0,0 +1,406 @@
+      times_df ::= dt
+      // scalastyle:off
+      print("Dataframes", times_df)
+      // scalastyle:on
```
The print statement should move out from the loop.

WeichenXu123 commented:
As the dataframe version is much slower than RDD version (currently test against vector of size 1)
I also guess there is some performance issue in `ObjectAggregationIterator.processInput()`
in the following code block:
```
  private def processInputs(): Unit = {
    // ...
    if (groupingExpressions.isEmpty) {
      // If there is no grouping expressions, we can just reuse the same buffer over and over again.
      val groupingKey = groupingProjection.apply(null)
      val buffer: InternalRow = getAggregationBufferByKey(hashMap, groupingKey)
      while (inputRows.hasNext) {
        val newInput = safeProjection(inputRows.next()) 
        processRow(buffer, newInput)
      }
    }
    ...
```
This statement `val newInput = safeProjection(inputRows.next())` maybe do some redundant data copy (for walk-around some bugs?)

cc @cloud-fan @liancheng

liancheng commented:
@WeichenXu123 and I did some profiling using `jvisualvm` and found that 40% of the time is spent in the copy performed by [this `safeProjection`][1]. This is a known issue used to fight against the false sharing issue @cloud-fan and I hit before.

@cloud-fan tried to fix this issue in #15082 but that PR didn't work out due to some other concerns (I can't remember all the details now).

@cloud-fan, any ideas about improving `ObjectHashAggregateExec` (e.g. adding code generation support)?

[1]: https://github.com/apache/spark/blob/03367d7aa3acd3abbcacba57ea75c8efa2a9a794/sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/ObjectAggregationIterator.scala#L154

cloud-fan commented:
The copy problem is fixed in https://github.com/apache/spark/pull/18483 , I think we can remove this workaround in `ObjectHashAggregateExec`.

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala:
```diff
@@ -0,0 +1,799 @@
+  }
+
+  def structureForMetrics(metrics: Seq[Metrics]): StructType = {
+    val dct = allMetrics.map { case (n, m, dt, _) => m -> (n, dt) }.toMap
```
`m -> (n, dt)` This expr has some syntax problem.
The scala complier will turn `->` into `.->` method and will treat `n, dt` as two parameters.
And this cause compiling error (after rebase this PR with master)
we can use:
`m -> (n -> dt)` or `(m, (n, dt))` or `m -> ((n, dt))` 
instead.

thunterdb commented:
I am going to close this PR, since this is being taken over by @WeichenXu123  in #18798 .
